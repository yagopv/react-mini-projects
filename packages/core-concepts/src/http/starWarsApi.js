import axios from "axios";

export async function getStarWarsPeople(url) {
  return axios.get(url);
}
