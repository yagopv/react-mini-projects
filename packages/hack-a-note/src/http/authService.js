import axios from "axios";

export function login(email, password) {
  return axios.post(`${process.env.REACT_APP_BACKEND_URL}/auth`, null, {
    auth: {
      username: email,
      password
    }
  });
}

export function register({ name, email, password }) {
  return axios.post(`${process.env.REACT_APP_BACKEND_URL}/users`, {
    name,
    email,
    password
  });
}
