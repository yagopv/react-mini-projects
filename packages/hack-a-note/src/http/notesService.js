import axios from "axios";

export function getNotes() {
  return axios.get(`${process.env.REACT_APP_BACKEND_URL}/notes`);
}

export function updateNote(note) {
  return axios.put(
    `${process.env.REACT_APP_BACKEND_URL}/notes/${note.id}`,
    note
  );
}

export function createNote(note) {
  return axios.post(`${process.env.REACT_APP_BACKEND_URL}/notes`, note);
}

export function deleteNote(noteId) {
  return axios.delete(`${process.env.REACT_APP_BACKEND_URL}/notes/${noteId}`);
}
