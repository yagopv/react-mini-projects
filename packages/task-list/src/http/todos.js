import axios from 'axios';

const BASE_URL = 'https://5dda901c5730550014fe78e2.mockapi.io/api/v1';

export function getTodos() {
  return axios.get(`${BASE_URL}/todo`);
}

export function addTodo(todo) {
  return axios.post(`${BASE_URL}/todo`, todo);
}

export function removeTodo(id) {
  return axios.delete(`${BASE_URL}/todo/${id}`);
}

export function updateTodo(id, todo) {
  return axios.put(`${BASE_URL}/todo/${id}`, todo);
}
