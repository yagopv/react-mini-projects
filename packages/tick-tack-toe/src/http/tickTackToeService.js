import axios from 'axios';

const BASE_URL =
  'https://5dda901c5730550014fe78e2.mockapi.io/api/v1/tick-tack-toe';

export function getLatestGames() {
  return axios.get(`${BASE_URL}?page=1&limit=20`);
}

export function addGame({ player1, player2, winner, type }) {
  return axios.post(`${BASE_URL}`, {
    player1,
    player2,
    winner,
    type
  });
}
