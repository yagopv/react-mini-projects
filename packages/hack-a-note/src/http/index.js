import axios from "axios";
import { login, register } from "./authService";
import { getNotes } from "./notesService";

// Utility function para comprobar si la URL necesita un Bearer o no
function isBearerTokenRequired(url) {
  const parsedURL = new URL(url);
  if (["/auth", "/user"].includes(parsedURL.pathname)) {
    return false;
  }
  return true;
}

// Al iniciar este módulo recuperamos el usuario guardado del localStorage.
// La idea es la misma que en el auth-context.js. Si el usuario refresca la página
// mantengo el estado en el que estaba
const currentUser = JSON.parse(localStorage.getItem("currentUser"));

// Iniciamos la variable token con lo almacenado en el localStorage
// Si no tenia nada la inicio a null
// La siguiente condicion de OR si la primera parte es true entonces se queda con el
// resultado de la primera condición y en caso contrario con la segunda (null)
let token = (currentUser && currentUser.token) || null;

// Definimos interceptors de request y response
// REQUEST
// -------
// En cada una de las peticiones se ejecutaran estas funciones
// Si la cadena de interceptors no falla entonces se irá ejecutando la
// primera función, si falla alguno de los interceptors entonces irá por la segunda.
axios.interceptors.request.use(
  function(config) {
    // Compruebo si tengo token y si necesito enviarlo en el
    // header de Authorization
    if (token && isBearerTokenRequired(config.url)) {
      config.headers["Authorization"] = `Bearer ${token}`;
    }
    // Acordarnos de devolver la config!!. Si no no sigue la cadena
    // y la petición no se hace
    return config;
  },
  function(error) {
    // Siempre devolver el error de esta forma, a través de Promise.reject
    return Promise.reject(error);
  }
);

// RESPONSE
// --------
// Cada vez que llegue una respuesta entonces se ejecutarán los interceptors
// La primera función si la respuesta es correcta
// La segunda si la respuesta trae un status >=400 o bien si tengo mas interceptors
// y alguno produce un error
axios.interceptors.response.use(
  function(response) {
    // Si la respuesta trae un token entonces lo almaceno
    // Mi aplicación supone que mi backend envia un objeto con { token, user } siempre
    // que el usuario se identifica (Login, Registro)
    // En otra aplicación podría ser diferente
    if (response.data.token) {
      localStorage.setItem("currentUser", JSON.stringify(response.data));
      token = response.data.token;
    }
    return response;
  },
  function(error) {
    // En caso de que el token expire (401)
    // y no sea el endpoint de login (que tambien devuelve 401 cuando las credenciales son invalidas)
    // Entonces redirijo a la URL de login y limpio el localStorage
    if (
      error.response.status === 401 &&
      error.config.url.indexOf("/auth") === -1
    ) {
      localStorage.removeItem("currentUser");
      window.location.href = "/login";
    }
    // Siempre devolver el error de esta forma, a través de Promise.reject
    return Promise.reject(error);
  }
);

export { login, register, getNotes };
