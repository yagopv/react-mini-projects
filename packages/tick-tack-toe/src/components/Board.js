import React, { useState, useEffect } from 'react';
import { Square } from './Square';
import { calculateWinner } from '../lib/utils';
import { Status } from './Status';
import { addGame } from '../http/tickTackToeService';

export function Board({
  player1,
  player2,
  onShowScoreBoardChange,
  onPlayersChanged
}) {
  const [squares, setSquares] = useState(Array(9).fill(null));
  const [isPlayer1Playing, setIsPlayer1Playing] = useState(true);
  const [winner, setWinner] = useState(null);

  const handleClick = i => {
    const sqCopy = [...squares];

    // Si ya ha sido pulsado o ya hay ganador
    if (sqCopy[i] || calculateWinner(squares)) {
      return;
    }

    // Quien esta jugando? X o Y
    sqCopy[i] = isPlayer1Playing ? 'X' : 'O';

    // Cambiar estado
    setSquares(sqCopy);
    setIsPlayer1Playing(!isPlayer1Playing);
  };

  useEffect(() => {
    const w = calculateWinner(squares);

    if (w === 'X') {
      setWinner(player1);
    }
    if (w === 'O') {
      setWinner(player2);
    }
    if (w) {
      addGame({
        player1,
        player2,
        winner: w === 'X' ? player1 : player2,
        type: w
      }).then(() => {
        onShowScoreBoardChange(true);
      });
    }
  }, [squares, player1, player2, onShowScoreBoardChange]);

  return (
    <React.Fragment>
      <Status
        rounds={squares.filter(s => s !== null).length}
        winner={winner}
        next={isPlayer1Playing ? `${player1} (X)` : `${player2} (O)`}
      />
      <div className="grid">
        {squares.map((square, i) => {
          return (
            <Square
              key={i}
              value={square}
              onChangeSquare={() => handleClick(i)}
            />
          );
        })}
      </div>
      <div className="buttons">
        <button
          className="btn"
          onClick={() => {
            setSquares(Array(9).fill(null));
            setIsPlayer1Playing(true);
            setWinner(null);
            onShowScoreBoardChange(false);
          }}
        >
          Reset
        </button>
        <button
          className="btn"
          onClick={() => {
            onPlayersChanged([]);
          }}
        >
          Enter Players
        </button>
      </div>
    </React.Fragment>
  );
}
