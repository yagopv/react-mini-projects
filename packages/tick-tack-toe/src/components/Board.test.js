import React from 'react';
import { render, fireEvent, wait } from '@testing-library/react';
import { Board } from './Board';
import * as http from '../http/tickTackToeService';

describe('<Board />', () => {
  it('should allow to complete the game', async () => {
    const onShowScoreBoardChange = jest.fn();
    http.addGame = jest.fn().mockResolvedValueOnce({ status: 201 });

    const { getAllByTestId, getByText, container } = render(
      <Board
        player1="p1"
        player2="p2"
        onShowScoreBoardChange={onShowScoreBoardChange}
      />
    );

    expect(getByText(/next player: p1/i)).toBeDefined();
    expect(container).toMatchSnapshot();

    const squares = getAllByTestId('square');
    fireEvent.click(squares[0]);
    expect(getByText(/next player: p2/i)).toBeDefined();
    expect(getByText('X')).toBeDefined();
    fireEvent.click(squares[1]);
    expect(getByText('O')).toBeDefined();
    fireEvent.click(squares[4]);
    fireEvent.click(squares[5]);
    fireEvent.click(squares[8]);

    expect(getByText(/winner: p1/i));

    await wait(() => {
      expect(onShowScoreBoardChange).toHaveBeenCalled();
      expect(onShowScoreBoardChange).toHaveBeenCalledTimes(1);
      expect(onShowScoreBoardChange).toHaveBeenCalledWith(true);
    });
  });
});
