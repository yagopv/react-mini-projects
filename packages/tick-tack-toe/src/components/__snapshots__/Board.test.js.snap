// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Board /> should allow to complete the game 1`] = `
<div>
  <div
    class="status"
    data-testid="status"
  >
    Next player: p1 (X)
  </div>
  <div
    class="grid"
  >
    <button
      class="square"
      data-testid="square"
    />
    <button
      class="square"
      data-testid="square"
    />
    <button
      class="square"
      data-testid="square"
    />
    <button
      class="square"
      data-testid="square"
    />
    <button
      class="square"
      data-testid="square"
    />
    <button
      class="square"
      data-testid="square"
    />
    <button
      class="square"
      data-testid="square"
    />
    <button
      class="square"
      data-testid="square"
    />
    <button
      class="square"
      data-testid="square"
    />
  </div>
  <div
    class="buttons"
  >
    <button
      class="btn"
    >
      Reset
    </button>
    <button
      class="btn"
    >
      Enter Players
    </button>
  </div>
</div>
`;
